### ISSUE TITLE
Investigate interaction between Andras Kovacs' implicit-fun-elaboration + Matthias Blume's polymorphic row types

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4423]
This task is automatically imported from the old Task Issue Board and it was originally created by Edward Kmett.
Original issue is [here](https://www.pivotaltracker.com/story/show/182495855).
<hr />

Performing initial work on what an actionable type system for Enso would look like. 

The short-term goal is to see if row types (as in MLPolyR) can play well with either Andras Kovacs' approach for `implicit-fun-elaboration`, or if we'd need to move to something heavier, like the "Quick Look" machinery used by GHC 9+.

https://people.cs.uchicago.edu/~blume/papers/icfp06.pdf
https://icfp20.sigplan.org/details/icfp-2020-papers/9/Elaboration-with-First-Class-Implicit-Function-Types

The primarily deliverable for this period is the answer to the question, though a sketch of a type-checker (possibly without surface syntax) will be built up in the meantime. This will inform how much of a type system I think we can have

## Comments:
**Edward Kmett** reports a new **STANDUP** for today (2022-06-16):

**Progress:** Spent the night talking with Andras Kovacs about some possible interactions between his type system which uses a limited notion of rows and a proper row type system. It should be finished by 2022-06-30.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/182495855). Digging through the Quick Look patch for GHC to see how bad it is to just instantiate monotypes spuriously as polytypes. (Enso Bot - Jun 17, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-06-17):

**Progress:** Mined through my old "ulf" codebase to see what pieces would be usable to use as a platform for testing. It should be finished by 2022-06-30.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/182495855). Much the same, though I'll try to dig into the Quick Look paper and source as I'd previously intended for today. (Enso Bot - Jun 17, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-06-20):

**Progress:** Working my way through the Quick Look patches. It should be finished by 2022-06-30.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/182495855). Still working through QuickLook. It does confusing things with polytypes. (Enso Bot - Jun 20, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for yesterday (2022-06-21):

**Progress:** Finally parsed what I think is the core QuickLook trick. It should be finished by 2022-06-30.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/182495855). Pushing around greek symbols. Meeting with Jaroslav. (Enso Bot - Jun 22, 2022)
<hr />
**Edward Kmett** reports a new **STANDUP** for today (2022-06-28):

**Progress:** Trying out "boxy types" as a path forward. Getting back on the standup wagon, after my previous postings crashed Enso Bot 1.0, now that we have its new, beefier replacement. It should be finished by 2022-06-30.

**Next Day:** Next day I will be working on the [#same task](https://www.pivotaltracker.com/story/show/182495855). Starting a toy boxy typechecker. After that I'll go through and re-evaluate how long this phase of figuring out the typechecker will take. (Enso Bot - Jun 29, 2022)
<hr />
The short version of this update is that they didn't play well together. Instantiation of row types at polytypes requires a mixed prefix of quantifiers forall ... exists ... forall ..., but when I go to apply the implicit-fun-elaboration story to that mix of quantifiers I have to distribute the inner foralls out over the existentials. This might be possible with some more advanced notion of a 'telescope' but it is currently beyond my ken. (Edward Kmett - Jul 11, 2022)
<hr />
